File: frontend/agentplatform/src/app/services/message.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Message {
  id: number;
  runId: number;
  senderAgentId?: number;
  recipientAgentId?: number;
  messageType: string; // standard, system, delegation, approval_request, etc.
  subject?: string;
  content: string;
  metadata?: string;
  status: string; // sent, read, archived, deleted
  createdAt: string;
  updatedAt: string;
}

export interface MessageRequest {
  runId?: number;
  senderAgentId?: number;
  recipientAgentId?: number;
  messageType?: string;
  subject?: string;
  content: string;
  metadata?: string;
}

@Injectable({
  providedIn: 'root'
})
export class MessageService {
  private apiUrl = '/api/runs';

  constructor(private http: HttpClient) { }

  /**
   * Send a new message
   */
  sendMessage(runId: number, request: MessageRequest): Observable<Message> {
    return this.http.post<Message>(`${this.apiUrl}/${runId}/messages`, request);
  }

  /**
   * Get all messages for a run
   */
  getRunMessages(runId: number): Observable<Message[]> {
    return this.http.get<Message[]>(`${this.apiUrl}/${runId}/messages`);
  }

  /**
   * Get paginated messages for a run
   */
  getRunMessagesPaginated(runId: number, page: number = 0, size: number = 20): Observable<any> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/${runId}/messages/paginated`, { params });
  }

  /**
   * Get a specific message
   */
  getMessage(runId: number, messageId: number): Observable<Message> {
    return this.http.get<Message>(`${this.apiUrl}/${runId}/messages/${messageId}`);
  }

  /**
   * Get messages sent by a specific agent
   */
  getMessagesBySender(runId: number, agentId: number): Observable<Message[]> {
    return this.http.get<Message[]>(`${this.apiUrl}/${runId}/messages/sender/${agentId}`);
  }

  /**
   * Get messages received by a specific agent
   */
  getMessagesForRecipient(runId: number, agentId: number): Observable<Message[]> {
    return this.http.get<Message[]>(`${this.apiUrl}/${runId}/messages/recipient/${agentId}`);
  }

  /**
   * Get messages of a specific type
   */
  getMessagesByType(runId: number, messageType: string): Observable<Message[]> {
    return this.http.get<Message[]>(`${this.apiUrl}/${runId}/messages/type/${messageType}`);
  }

  /**
   * Get unread messages for an agent
   */
  getUnreadMessages(runId: number, agentId: number): Observable<Message[]> {
    return this.http.get<Message[]>(`${this.apiUrl}/${runId}/messages/unread/${agentId}`);
  }

  /**
   * Mark message as read
   */
  markAsRead(runId: number, messageId: number): Observable<Message> {
    return this.http.put<Message>(`${this.apiUrl}/${runId}/messages/${messageId}/read`, {});
  }

  /**
   * Delete a message
   */
  deleteMessage(runId: number, messageId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${runId}/messages/${messageId}`);
  }

  /**
   * Get message count for a run
   */
  getMessageCount(runId: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/${runId}/messages/stats/count`);
  }

  /**
   * Get unread message count for an agent
   */
  getUnreadMessageCount(runId: number, agentId: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/${runId}/messages/unread-count/${agentId}`);
  }
}

File: backend/src/main/java/com/agentplatform/backend/runs/MessageService.java
package com.agentplatform.backend.runs;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class MessageService {

    private final MessageRepository messageRepository;

    public MessageService(MessageRepository messageRepository) {
        this.messageRepository = messageRepository;
    }

    /**
     * Send a new message
     */
    public MessageResponse sendMessage(MessageRequest request) {
        if (request.getRunId() == null) {
            throw new IllegalArgumentException("Run ID is required");
        }
        if (request.getContent() == null || request.getContent().isEmpty()) {
            throw new IllegalArgumentException("Message content is required");
        }

        Message message = new Message(request.getRunId(), request.getContent());
        message.setSenderAgentId(request.getSenderAgentId());
        message.setRecipientAgentId(request.getRecipientAgentId());

        if (request.getMessageType() != null) {
            message.setMessageType(request.getMessageType());
        }
        if (request.getSubject() != null) {
            message.setSubject(request.getSubject());
        }
        if (request.getMetadata() != null) {
            message.setMetadata(request.getMetadata());
        }

        Message saved = messageRepository.save(message);
        return new MessageResponse(saved);
    }

    /**
     * Get a message by ID
     */
    public Optional<MessageResponse> getMessageById(Long id) {
        return messageRepository.findById(id)
                .map(MessageResponse::new);
    }

    /**
     * Get all messages for a run
     */
    public List<MessageResponse> getRunMessages(Long runId) {
        return messageRepository.findByRunIdOrderByCreatedAtAsc(runId)
                .stream()
                .map(MessageResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Get paginated messages for a run
     */
    public Page<MessageResponse> getRunMessagesPaginated(Long runId, Pageable pageable) {
        return messageRepository.findByRunId(runId, pageable)
                .map(MessageResponse::new);
    }

    /**
     * Get messages sent by a specific agent
     */
    public List<MessageResponse> getMessagesBySender(Long runId, Long agentId) {
        return messageRepository.findBySenderAgentIdAndRunId(agentId, runId)
                .stream()
                .map(MessageResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Get messages received by a specific agent
     */
    public List<MessageResponse> getMessagesForRecipient(Long runId, Long agentId) {
        return messageRepository.findByRecipientAgentIdAndRunId(agentId, runId)
                .stream()
                .map(MessageResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Get messages of a specific type
     */
    public List<MessageResponse> getMessagesByType(Long runId, String messageType) {
        return messageRepository.findByRunIdAndMessageType(runId, messageType)
                .stream()
                .map(MessageResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Get unread messages for an agent
     */
    public List<MessageResponse> getUnreadMessages(Long runId, Long agentId) {
        return messageRepository.findUnreadMessagesForAgent(runId, agentId)
                .stream()
                .map(MessageResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Mark a message as read
     */
    public MessageResponse markMessageAsRead(Long messageId) {
        Message message = messageRepository.findById(messageId)
                .orElseThrow(() -> new IllegalArgumentException("Message not found"));

        message.setStatus("read");
        message.setUpdatedAt(Instant.now());

        Message updated = messageRepository.save(message);
        return new MessageResponse(updated);
    }

    /**
     * Delete a message
     */
    public void deleteMessage(Long messageId) {
        messageRepository.deleteById(messageId);
    }

    /**
     * Count messages in a run
     */
    public long countMessages(Long runId) {
        return messageRepository.countByRunId(runId);
    }

    /**
     * Count unread messages for an agent
     */
    public long countUnreadMessages(Long runId, Long agentId) {
        return messageRepository.countByRunIdAndRecipientAgentIdAndStatus(runId, agentId, "sent");
    }
}
